[Refactor] simplify multimodal data processing

JustinTong0323 opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

This pull request aims to refactor and simplify the multimodal data processing within the system. The core change involves centralizing model-specific data into a single dictionary within the MultimodalDataItem class, reducing boilerplate and improving extensibility. This refactoring impacts how various multimodal models and their processors handle and access their specific input features.

Highlights

- Centralized Multimodal Data Storage: The MultimodalDataItem class has been refactored to consolidate model-specific attributes (e.g., image_grid_thw, audio_feature_lens) into a new model_specific_data dictionary, simplifying the data structure and making it more generic.
- Streamlined Multimodal Processors: Multimodal processors (e.g., CLIP, Deepseek-VL, Mllama) have been updated to utilize the new model_specific_data structure, reducing redundant code and improving consistency in how data is prepared for models.
- Model Adaptations: Numerous model implementations (e.g., Deepseek-VL2, Gemma3N, Kimi-VL, MiniCPM, Phi4MM, Qwen2.5-VL, Qwen2-Audio, Qwen2-VL) have been adjusted to retrieve their specific input features from the model_specific_data dictionary.
- Processor Utility Enhancements: The BaseMultimodalProcessor now includes improved logic for collecting and combining multimodal data, leveraging the new centralized data structure.
- File Renaming/Restructuring: The qwen_audio.py processor file has been moved and renamed, indicating a potential restructuring of the multimodal processor directory.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

JustinTong0323 committed:
refactor: simplify multimodal data processing

This commit refactors how multimodal data is handled, promoting a more organized and maintainable structure.

- Consolidates model-specific data within a dictionary inside MultimodalDataItem, reducing code duplication and improving data organization.
- Removes the qwen_audio.py processor as part of the refactoring.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

mickqian commented:
this has quite some duplication with #7924. Can we prioritize one?

JustinTong0323 committed:
revert modification for mllama4

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
fix lint

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Merge branch 'main' into feat-simlify-multimodal-dataitem-processing

JustinTong0323 committed:
refactor: update type hints for model_specific_data and items in schedule_batch.py and base_processor.py

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Refactors how model-specific multimodal data is handled.

Moves model-specific data from individual `MultimodalDataItem` attributes
to a dictionary within `model_specific_data` for better organization
and maintainability. This change simplifies access to model-specific
features during processing.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-simlify-multimodal-dataitem-processing

JustinTong0323 commented:
/gemini review

JustinTong0323 committed:
fix: restore change fot models/mllama.py

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix: unify multimodal token handling across processors

Updated multiple image processor classes to use a shared `mm_tokens` attribute for multimodal token management, improving consistency and reducing redundancy in the codebase. This change enhances maintainability and simplifies the processing of multimodal data.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
misc: enhance type hints in base_processor.py

Updated type hints for the `data_iterators` parameter and `all_collected_items` variable in the BaseMultimodalProcessor class to improve code clarity and type safety. This change aids in better understanding and maintainability of the multimodal processing logic.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 commented:
/gemini summary

JustinTong0323 committed:
fix: use get method for model_specific_data access in multiple models

Updated various model files to replace direct dictionary access with the get method for retrieving values from model_specific_data. This change enhances code robustness by preventing potential KeyErrors and improves consistency across the codebase.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Removes `max_req_input_len` from multimodal processor calls.

This simplifies the processor interfaces and removes a parameter
that is no longer needed in the processing logic.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-simlify-multimodal-dataitem-processing

JustinTong0323 committed:
fix after merge main:
Moves multimodal data attributes to the `model_specific_data` dictionary within `MultimodalDataItem`.

This change improves code organization and flexibility by encapsulating model-specific data, such as image sizes and attention masks, within a dedicated dictionary. It also prepares the codebase for easier extension to support diverse multimodal models with varying data requirements.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Refactors audio processing in Qwen2

Refactors the audio processing logic in Qwen2 to improve efficiency and readability.

- Pre-collects and stores special token IDs to avoid redundant lookups.
- Extracts and consolidates multimodal token information.
- Simplifies the data processing flow.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Refactors multimodal token handling

Updates multimodal processors to correctly utilize and pass the image token ID.

This change ensures that the `image_token_id` is consistently accessed from the `mm_tokens` object instead of directly from the processor, leading to more reliable and maintainable code. It also initializes image_token_id in the MultimodalSpecialTokens for processors that need it.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Rename `precomputed_features` to `precomputed_embeddings`

This update changes references from `precomputed_features` to `precomputed_embeddings` across various modules, including the VLM input format, multimodal data item, and processor classes. This refactor enhances consistency in the handling of multimodal data and improves code clarity.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Remove `image_sizes` as common MultimodalDataItem attributes

This update removes the `image_sizes` attribute from the `MultimodalDataItem` class and clarifies the purpose of `precomputed_embeddings`. These changes enhance the organization of multimodal data handling and improve code clarity.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix: deepseek vl related issue

- Added `items` method to `DictOutput` class for better dictionary-like behavior.
- Renamed `images` attribute to `pixel_values` in `VLChatProcessorOutput` for consistency.
- Updated `DeepseekVL2ForCausalLM` to use `images_spatial_crop` instead of `image_spatial_crop`.
- Modified `DeepseekVL2ImageProcessor` to include `max_req_input_len` in the processing method.
- Adjusted multimodal processor calls to pass additional parameters for improved flexibility.

These changes streamline the multimodal data processing and enhance the overall organization of the codebase.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

mickqian commented on python/sglang/srt/managers/schedule_batch.py:
```diff
@@ -211,37 +211,11 @@ class MultimodalDataItem:
 
-    image_sizes: Tuple[int, int] = None
+    # the precomputed embeddings for the modality, e.g. image_emb for image, audio_emb for audio
+    precomputed_embeddings: Optional[Union[torch.Tensor, np.ndarray]] = None
```
maybe plural is not needed?

JustinTong0323 replied:
I'm not sure, as the embedding could represent multiple images, I think remain plural here is more general

mickqian commented on python/sglang/srt/multimodal/processors/janus_pro.py:
```diff
@@ -11,52 +11,35 @@
         image_data: List[Union[str, bytes]],
         input_text,
         request_obj,
-        max_req_input_len,
```
We might still need this, which is useful for controlling input len

JustinTong0323 replied:
Currently only deepseek vl2 use it and I remain as a optional kwarg.

mickqian commented on python/sglang/srt/multimodal/processors/mlama.py:
```diff
@@ -1,34 +1,38 @@
         super().__init__(hf_config, server_args, _processor)
+        self.IM_TOKEN = self._processor.image_token
+        self.IM_TOKEN_ID = self._processor.image_token_id
+        self.mm_tokens = MultimodalSpecialTokens(
```
nit: I would prefer skip storing `self.IM_TOKEN_ID`  itself. a variable in `mm_tokens` is enough, and it's easier if we want to return mm_tokens in the future

JustinTong0323 commented on python/sglang/srt/managers/schedule_batch.py:
```diff
@@ -306,8 +269,7 @@ def from_dict(obj: dict):
     def merge(self, other):
         self.feature += other.feature
-        self.image_sizes += other.image_sizes
-        self.image_offsets += other.image_offsets
+        self.offsets += other.offsets
         self.hash = hash((self.hash, other.hash))
         self.set_pad_value()
```
@mickqian What do you think about it? Maybe could resolve it when merge is used?

mickqian replied:
Do we have the merge logic for these field before? if not, it is not urgent

mickqian commented on python/sglang/srt/models/llava.py:
```diff
@@ -753,7 +759,9 @@ def get_image_feature(self, items: List[MultimodalDataItem]) -> torch.Tensor:
         for item in items:
             # in each item, we assume pixel_values is always batched
-            pixel_values, image_sizes = item.feature, item.image_sizes
+            pixel_values, image_sizes = item.feature, item.model_specific_data.get(
```
can we add a `get` method to `MultimodalDataItem`?

JustinTong0323 replied:
DONE

mickqian replied:
nit: sorry for letting you implement `get` earlier, but do you think this way it can be simpler? :
```python
class Item:
    def __getattr__(self, name):
        try:
            return self._data[name]
        except KeyError:
            raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")

# instead of `.get("some_fields")`
Item().some_fields
```

JustinTong0323 committed:
Refactors MultimodalDataItem attr access and usage

Refactors how multimodal models access data within `MultimodalDataItem`.

- Adds `__getitem__`, `get`, `__setitem__`, and `set` methods to `MultimodalDataItem` for more intuitive and flexible data access.
- Updates multimodal models to use these new methods instead of directly accessing `model_specific_data`.

This change improves code readability and maintainability by providing a consistent interface for accessing data associated with multimodal items.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
misc: remove redundant attr in MllamaImageProcessor

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix: update attribute access in MllamaForConditionalGeneration

Replaces direct attribute access with the new `get` method for `aspect_ratio_id` and `aspect_ratio_mask` in the `MllamaForConditionalGeneration` class. This change enhances code readability and aligns with recent refactoring efforts in the multimodal data handling.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix lint

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix llama atttr

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Temporarily disable MllamaServer tests due to instability

Commented out the TestMllamaServer class and its methods in test_vision_openai_server_a.py, indicating that Mllama is not stable for CI. This change prevents potential failures in the test suite while maintaining the code for future use.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix: qwen vl video attr error

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
fix lint

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-simlify-multimodal-dataitem-processing

JustinTong0323 committed:
Streamlines access to `MultimodalDataItem` attributes by using `__getattr__` for model-specific data.

This change improves code readability and conciseness in multimodal models like DeepseekVL2, Gemma3n, KimiVL, Llava, MiniCPMO, MiniCPMv, Mistral, Mllama, Phi4MM, Qwen2, and Qwen2Audio by allowing direct access to data attributes instead of using `item.get("key")`. It also fixes a bug in Phi4MMProcessorAdapter where the original hf_key was not being deleted from the result, leading to incorrect assignment.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into feat-simlify-multimodal-dataitem-processing

mickqian commented on test/srt/test_vision_openai_server_common.py:
```diff
@@ -308,19 +308,35 @@ def test_video_images_chat_completion(self):
+        ====================== video_response =====================
+        {video_response}
+        ===========================================================
+        should contain 'iPod' or 'device' or 'microphone'
```
better not hard code this

JustinTong0323 committed:
Merge branch 'main' into feat-simlify-multimodal-dataitem-processing
